src/modules/commander/state_machine_helper.cpp

```diff
@@ -580,25 +577,35 @@ bool set_nav_state(vehicle_status_s &status, actuator_armed_s &armed, commander_
 		} else if (is_armed && check_invalid_pos_nav_state(status, old_failsafe, mavlink_log_pub, status_flags, false, true)) {
 			// nothing to do - everything done in check_invalid_pos_nav_state
 		} else if (status.data_link_lost && data_link_loss_act_configured && !landed && is_armed) {
-			/* also go into failsafe if just datalink is lost, and we're actually in air */
-			set_link_loss_nav_state(status, armed, status_flags, internal_state, data_link_loss_act, 0);
-
+			// Data link lost, data link loss reaction configured -> do configured reaction
 			enable_failsafe(status, old_failsafe, mavlink_log_pub, reason_no_datalink);
+			set_link_loss_nav_state(status, armed, status_flags, internal_state, data_link_loss_act, 0);
 
-		} else if (rc_lost && !data_link_loss_act_configured && status.data_link_lost && is_armed) {
-			/* go into failsafe if RC is lost and datalink is lost and datalink loss is not set up */
+		} else if (status.rc_signal_lost && !(param_com_rcl_except & RCLossExceptionBits::RCL_EXCEPT_LOITER)
+			   && status_flags.rc_signal_found_once && is_armed && !landed) {
+			// RC link lost, rc loss not disabled in loiter, RC was used before -> RC loss reaction after delay
+			// Safety pilot expects to be able to take over by RC in case anything unexpected happens
 			enable_failsafe(status, old_failsafe, mavlink_log_pub, reason_no_rc);
+			set_link_loss_nav_state(status, armed, status_flags, internal_state, rc_loss_act, param_com_rcl_act_t);
 
+		} else if (status.rc_signal_lost && !(param_com_rcl_except & RCLossExceptionBits::RCL_EXCEPT_LOITER)
+			   && status.data_link_lost && !data_link_loss_act_configured
+			   && is_armed && !landed) {
+			// All links lost, no data link loss reaction configured -> immediately do RC loss reaction
+			// Lost all communication, by default it's considered unsafe to continue the mission
+			// This is only reached when flying mission completely without RC (it was not present since boot)
+			enable_failsafe(status, old_failsafe, mavlink_log_pub, reason_no_datalink);
 			set_link_loss_nav_state(status, armed, status_flags, internal_state, rc_loss_act, 0);
 
-		} else if (status.rc_signal_lost) {
-			/* don't bother if RC is lost if datalink is connected */
-
-			/* this mode is ok, we don't need RC for LOITERing */
-			status.nav_state = vehicle_status_s::NAVIGATION_STATE_AUTO_LOITER;
+		} else if (status.rc_signal_lost && (param_com_rcl_except & RCLossExceptionBits::RCL_EXCEPT_LOITER)
+			   && status.data_link_lost && !data_link_loss_act_configured
+			   && is_armed && !landed) {
+			// All links lost, all link loss reactions disabled -> return
+			// Pilot disabled all reactions, return to avoid lost vehicle
```

Comment:
So that's now completely a "failsafe" for users but will be surprising for someone that actually wants that.

Reply:
So the use case would be to let the vehicle loiter without any link right? Is this more a simulation case or do you expect actual use cases e.g. switching from one link to another while loitering?

In any case the concern is valid. The reason I added it was because I expected a mission to command the vehicle to loiter for a certain time is enough but maybe it's not. Should we then remove the case and just react on losing either link and the user is able to disable the reaction (data link being disabled by default?).

I'd like this solution more if data link loss reaction was enabled by default but would only react on actual loss and not on never having any data link essentially making it a requirement to exist in the first place.

Reply:
Ah, so this is only for loiter but not for mission?

Reply:
I'll change it back for now and we'll improve e.g. with the default data link loss next iteration. After all this was not really the scope of this pr but just added because I had concerns when users might still accidentally disable all failsafes. Let's go in steps.

Reply:
It's only for loiter and takeoff. Mission finishes before returning when all links are lost.
I removed it now in https://github.com/PX4/PX4-Autopilot/pull/17785/commits/207b5705b94b4c95a0c94f1e9dafc543823c3f45 mainly because there could be SITL cases where you want to fly with `commander takeoff` and no link which would not be possible anymore afterwards. I don't see another real-world use case but maybe it exists. We can add the data link loss by default and then the user needs to explicitly disable both at which point I'd say he hopefully knows what he does.